pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -42,6 +57,360 @@ type Adapter struct {
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to watch deployment")
+	}
+	defer w.Stop()
+	success := make(chan *unstructured.Unstructured)
+	failure := make(chan error)
+
+	go func() {
+		defer close(success)
+		defer close(failure)
+
+		for {
+			val, ok := <-w.ResultChan()
+			if !ok {
+				failure <- errors.New("watch channel was closed")
+				return
+			}
+			if watchObject, ok := val.Object.(*unstructured.Unstructured); ok {
+				// TODO: Add more details on what to check to see if object deployment is complete
+				// Currently only checks to see if status.conditions[] contains a condition with type = conditionTypeValue
+				condition := getNamedCondition(watchObject, conditionTypeValue)
+				if condition != nil {
+					if condition["status"] == "Fail" {
+						failure <- fmt.Errorf("manifest deployment %s failed", applicationName)
+					} else if condition["status"] == "True" {
+						success <- watchObject
+					}
+				}
+			}
+		}
+	}()
+
+	select {
+	case val := <-success:
+		return val, nil
+	case err := <-failure:
+		return nil, err
+	case <-time.After(30 * time.Second):
+		return nil, errors.Errorf("timeout while waiting for %s manifest deployment completion", applicationName)
+	}
+}
+
+// Build image for devfile project
+func (a Adapter) Deploy(parameters common.DeployParameters) (err error) {
+	namespace := a.Client.Namespace
+	applicationName := a.ComponentName + "-deploy"
+	deploymentManifest := &unstructured.Unstructured{}
+
+	log.Info("\nDeploying manifest")
+
+	// Specify the substitution keys and values
+	yamlSubstitutions := map[string]string{
+		"CONTAINER_IMAGE": parameters.Tag,
+		"COMPONENT_NAME":  applicationName,
+		"PORT":            determinePort(parameters.EnvSpecificInfo),
+	}
+
+	// Build a yaml decoder with the unstructured Scheme
+	yamlDecoder := yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+
+	// This will override if manifest.yaml is present
+	writtenToManifest := false
+	manifestFile, err := os.Create(filepath.Join(a.Context, ".odo", "manifest.yaml"))
+	if err != nil {
+		err = manifestFile.Close()
+		return errors.Wrap(err, "Unable to create the local manifest file")
+	}
+
+	defer func() {
+		merr := manifestFile.Close()
+		if err == nil {
+			err = merr
+		}
+	}()
+
+	manifests := bytes.Split(parameters.ManifestSource, []byte("---"))
+	for _, manifest := range manifests {
+		if len(manifest) > 0 {
+			// Substitute the values in the manifest file
+			deployYaml := substitueYamlVariables(manifest, yamlSubstitutions)
+
+			_, gvk, err := yamlDecoder.Decode([]byte(deployYaml), nil, deploymentManifest)
+			if err != nil {
+				return errors.Wrap(err, "Failed to decode the manifest yaml")
+			}
+
+			deployJSON, err := deploymentManifest.MarshalJSON()
+			if err != nil {
+				return errors.Wrap(err, "Deployment manifest is invalid")
+			}
+			klog.V(3).Infof("Deploy manifest:\n\n%s\n", deployJSON)
```

Comment:
Need to document why we need this

Reply:
We can delete it as it was only for my testing / debug. 
Left it in test whether the manifest was valid.
If we leave the MarshalJSON method in, then we don't need to have the klog.V(3).Infof log output.

Reply:
Removed this section.